tests/integration/security/cert_mtls/cert_mtls_test.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certmtls
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/test/echo/common/scheme"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/echoboot"
+	"istio.io/istio/pkg/test/framework/components/environment/kube"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/framework/resource/environment"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/integration/security/util"
+	"istio.io/istio/tests/integration/security/util/cert"
+	"istio.io/istio/tests/integration/security/util/connection"
+)
+
+var (
+	testNamespace namespace.Instance
+	testCtx       framework.TestContext
+	testStruct    *testing.T
```

Comment:
We should not use global variables for this, its hard to understand

Reply:
The global variables are used because the function checkCACert() passed to retry.UntilSuccessOrFail() has a declaration of "fn func() error" as follows so the variables can only be passed through the global variables.

func UntilSuccessOrFail(t test.Failer, fn func() error, options ...Option)

Reply:
That does not require using global variable. You can provide a helper function that takes in the arguments needed

Reply:
About providing a helper function that takes in the arguments needed, is it having a calling flow like the following? Where do the arguments in the helper() come from? Are they coming from global variables? Can you provide more details (e.g., a calling flow) about providing a helper function that takes in the arguments needed?

func TestCertMtls(t *testing.T) {
   ....
   retry.UntilSuccessOrFail(t, checkCACert, retry.Delay(time.Second), retry.Timeout(10*time.Second))
   ....
}

func checkCACert() error {
   helper(testCtx, testStruct, other variables etc)
}

Reply:
```go
func TestCertMtls(t testing.T) {
....
checkCACert(t, ctx)
....
}

func checkCACert(t, ctx) error {
retry.UntilSuccessOrFail(t, func() error {
   ...do something...
})
}
```

Reply:
Or

```go
func TestCertMtls(t testing.T) {
....
retry.UntilSuccessOrFail(t, func() error {
return checkCaCert(t, ctx)
}, retry.Delay(time.Second), retry.Timeout(10time.Second))
....
}

func checkCACert(t, ctx) error {
helper(testCtx, testStruct, other variables etc)
}
```

Reply:
Thanks! The PR has been revised to remove the global variables.